Azure/sonic-buildimage
# Introduce sonic-py-common package

**- Why I did it**
To consolidate common SONiC Python-language functionality into one shared package and eliminate duplicate code. This is the first step toward resolving https://github.com/Azure/sonic-buildimage/issues/4999

The package currently includes three modules:
- daemon_base
- device_info
- logger

**- How I did it**
Add a new Python package, "sonic-py-common". This will replace the sonic-daemon-base package as well as the sonic_device_util.py file in sonic-config-engine.

I am building both Python 2 and Python 3 versions of the package. The Python 2 package is for backward-compatibility with existing Python 2 code, and the Python 3 version is for use in the SNMP container as well as to prepare for the eventual migration of all Python code to Python 3, since Python 2 is no longer supported as of January 1, 2020.

I think that `db_connect()` and `load_platform_util()` should also be moved out of daemon_base.py and into a separate module, but I opted against it for this PR, just to make the initial transition from sonic-daemon-base to sonic-py-common more straightforward. This can be taken care of in a future refactor.

**- How to verify it**
Test using the following script (replace `python2` with `python3` in shebang to test Python 3 package):
```python
#!/usr/bin/env python2

from sonic_py_common import device_info, logger, daemon_base

print("machine_info: {}".format(device_info.get_machine_info()))
print("asic_cfg_path: {}".format(device_info.get_asic_conf_file_path()))
print("platform: {}".format(device_info.get_platform()))
print("hwsku: {}".format(device_info.get_hwsku()))

plat, sku = device_info.get_platform_and_hwsku()
print("plat: {}, sku: {}".format(plat, sku))

plat_path, sku_path = device_info.get_paths_to_platform_and_hwsku_dirs()
print("plat_path: {}, sku_path: {}".format(plat_path, sku_path))

print("port_config_path: {}".format(device_info.get_path_to_port_config_file()))
print("version_info: {}".format(device_info.get_sonic_version_info()))
print("num_npus: {}".format(device_info.get_num_npus()))
print("is_multi_npu: {}".format(device_info.is_multi_npu()))
print("namespaces: {}".format(device_info.get_namespaces()))
print("all_namespaces: {}".format(device_info.get_all_namespaces()))

my_logger = logger.Logger("test")
my_logger.log_error("Test error message; Minimum log priority = PRIORITY_NOTICE")
my_logger.log_warning("Test warning message; Minimum log priority = PRIORITY_NOTICE")
my_logger.log_notice("Test notice message; Minimum log priority = PRIORITY_NOTICE")
# The following SHOULDN'T get logged
my_logger.log_info("Test info message; Minimum log priority = PRIORITY_NOTICE")
my_logger.log_debug("Test debug message; Minimum log priority = PRIORITY_NOTICE")

my_logger.set_min_log_priority_info()
my_logger.log_error("Test error message; Minimum log priority = PRIORITY_INFO")
my_logger.log_warning("Test warning message; Minimum log priority = PRIORITY_INFO")
my_logger.log_notice("Test notice message; Minimum log priority = PRIORITY_INFO")
my_logger.log_info("Test info message; Minimum log priority = PRIORITY_INFO")
# The following SHOULDN'T get logged
my_logger.log_debug("Test debug message; Minimum log priority = PRIORITY_INFO")

my_logger.set_min_log_priority_debug()
my_logger.log_error("Test error message; Minimum log priority = PRIORITY_DEBUG")
my_logger.log_warning("Test warning message; Minimum log priority = PRIORITY_DEBUG")
my_logger.log_notice("Test notice message; Minimum log priority = PRIORITY_DEBUG")
my_logger.log_info("Test info message; Minimum log priority = PRIORITY_DEBUG")
my_logger.log_debug("Test debug message; Minimum log priority = PRIORITY_DEBUG")


class MyDaemon(daemon_base.DaemonBase):
    def run(self):
        self.log_error("Test error message; Minimum log priority = PRIORITY_NOTICE")
        self.log_warning("Test warning message; Minimum log priority = PRIORITY_NOTICE")
        self.log_notice("Test notice message; Minimum log priority = PRIORITY_NOTICE")
        # The following SHOULDN'T get logged
        self.log_info("Test info message; Minimum log priority = PRIORITY_NOTICE")
        self.log_debug("Test debug message; Minimum log priority = PRIORITY_NOTICE")

        self.set_min_log_priority_info()
        self.log_error("Test error message; Minimum log priority = PRIORITY_INFO")
        self.log_warning("Test warning message; Minimum log priority = PRIORITY_INFO")
        self.log_notice("Test notice message; Minimum log priority = PRIORITY_INFO")
        self.log_info("Test info message; Minimum log priority = PRIORITY_INFO")
        # The following SHOULDN'T get logged
        self.log_debug("Test debug message; Minimum log priority = PRIORITY_INFO")

        self.set_min_log_priority_debug()
        self.log_error("Test error message; Minimum log priority = PRIORITY_DEBUG")
        self.log_warning("Test warning message; Minimum log priority = PRIORITY_DEBUG")
        self.log_notice("Test notice message; Minimum log priority = PRIORITY_DEBUG")
        self.log_info("Test info message; Minimum log priority = PRIORITY_DEBUG")
        self.log_debug("Test debug message; Minimum log priority = PRIORITY_DEBUG")

my_daemon = MyDaemon("test_daemon")
my_daemon.run()
```

**- Description for the changelog**
Consolidate common SONiC Python-language functionality into one shared package (sonic-py-common) and eliminate duplicate code.

**- Next Steps**
1. Migrate all Python applications in sonic-buildimage from sonic-daemon-base and sonic_device_util to sonic-py-common
2. Migrate all Python applications in submodules from sonic-daemon-base and sonic_device_util to sonic-py-common
3. Remove sonic-daemon-base and sonic_device_util entirely
4. Migrate common code from sonic-utilities into sonic-py-common
5. Add regression tests for sonic--py-common to sonic-mgmt repo
